### PR TITLE
Specify time-arrays discontinuities in diffrax solvers

### PR DESCRIPTION
**Pull request by pierreguilmin**
Fixes DYN-218.

Thanks to @xiasitan for spotting the bug. 😉

## MWE
```python
import dynamiqs as dq
import jax
import jax.numpy as jnp

# PWCTimeArray
times = [0, 1, 1.1, 2]
values = [0, 10, 0]
H = dq.pwc(times, values, -dq.sigmax())
result = dq.sesolve(H, dq.ground(), [0, 2])
print(result)
print(dq.expect(dq.sigmaz(), result.states[-1]).real)

# ModulatedTimeArray
f = lambda t: jax.lax.cond(
    jnp.logical_and(1 <= t, t < 1.1), lambda _: 10, lambda _: 0, None
)
H = dq.modulated(f, -dq.sigmax(), discontinuity_ts=[1.0, 1.1])
result = dq.sesolve(H, dq.ground(), [0, 2])
print(result)
print(dq.expect(dq.sigmaz(), result.states[-1]).real)
```

Before this PR (wrong result):
```
# 5 steps
# -1.0
# 5 steps
# -1.0
```

After this PR (correct result):
```
# 15 steps
# 0.4161431
# 15 steps
# 0.4161431
```

## Old comment
Note that this PR breaks the JIT. The reason is that `jnp.unique` is data-dependent, see doc here > https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.unique.html. We have three solutions:
1. Either `jump_ts` in Diffrax does not need an array with unique values, that would be ideal.
2. Otherwise we coud specify statically the `size` argument of `jnp.unique`, which is fine if the `H`, `jump_ops` are defined outside JIT, we can just feed `len(times)`.
3. Otherwise we could specify a default `size` to `jnp.unique` (specifically made for this purpose), but then we need to pick a `fill_value`, and we need to check what fits Diffrax.

Left to do to fully close the issue after we've fixed the JIT:
- [x] set `discontinuities_ts` to private or document in `TimeArray` API
- [x] pull `discontinuities_ts` up to the `modulated`/`timecallable` API
- [x] document this in the time-dependent tutorial

Note that for `modulated` _we could also_ automatically scan the function to detect discontinuities, but that's a bit niche, I think a first solution where we delegate to the user is good enough for now.

### CODE REVIEW

**Comment by pierreguilmin** [comment]
I'm also thinking, for the propagator solver to work with PWC pulse ([DYN-47](https://linear.app/dynamiqs/issue/DYN-47/use-pwc-hamiltonian-format-in-propagator-solvers)), we'll probably need this uniquely sorted time-array anyway.

**Comment by pierreguilmin** [comment]
Turns out the easy solution works, thanks @gautierronan for looking into this!
> 1. Either jump_ts in Diffrax does not need an array with unique values, that would be ideal.

File: dynamiqs/time_array.py [hunk 1]
```diff
@@ -167,6 +186,9 @@ class TimeArray(eqx.Module):
         mT _(TimeArray)_: Returns the time-array transposed over its last two
             dimensions.
         ndim _(int)_: Number of dimensions.
+        discontinuity_ts _(Array | None)_: Times at which there is a discontinuous jump
+            in the time-array values (the array is always sorted, but does not
+            necessarily contain unique values).
```
Comment (gautierronan) [inline dynamiqs/time_array.py:191]: Related to the main comment. I don't think the array is necesseraly sorted at this point.

**Review by gautierronan — CHANGES_REQUESTED** [review]
Just a single comment: I think we need to sort the `discontinuity_ts` array somewhere inside the `time_array.py` file. Right now, I believe we can pass an unsorted array and it will never be sorted if `sesolve` is used. For `mesolve`, it goes through `_concatenate_sort` so its fine.

The rest looks good to me!

**Comment by pierreguilmin** [comment]
> Just a single comment: I think we need to sort the discontinuity_ts array somewhere inside the time_array.py file.

Good catch!!

**Pull request merged**